src/SpinWaveTheory/SpinWaveTheory.jl

```diff
@@ -75,78 +77,79 @@ function generate_local_sun_gens(sys :: System)
-        Q̃_mat = Array{ComplexF64, 4}(undef, 2, 2, 5, Nₘ)
-
-        U_mat_2 = Matrix{ComplexF64}(undef, 2, 2)
-        U_mat_N = Matrix{ComplexF64}(undef, N, N)
-
-        for atom = 1:Nₘ
-            θ, ϕ = dipole_to_angles(sys.dipoles[1, 1, 1, atom])
-            U_mat_N[:] = exp(-1im * ϕ * s_mat_N[3]) * exp(-1im * θ * s_mat_N[2])
-            U_mat_2[:] = exp(-1im * ϕ * s_mat_2[3]) * exp(-1im * θ * s_mat_2[2])
-            @assert isapprox(U_mat_N * U_mat_N', I) "rotation matrix from (global frame to local frame) not unitary"
-            @assert isapprox(U_mat_2 * U_mat_2', I) "rotation matrix from (global frame to local frame) not unitary"
-            for μ = 1:3
-                s̃_mat[:, :, μ, atom] = Hermitian(U_mat_2' * s_mat_2[μ] * U_mat_2)
-            end
-            for ν = 1:5
-                Q̃_mat[:, :, ν, atom] = Hermitian(U_mat_N' * Q_mat_N[ν] * U_mat_N)[1:2, 1:2]
-            end
-
-            if !no_single_ion
-                T̃_mat[:, :, atom] = Hermitian(U_mat_N' * sys.interactions_union[atom].aniso.matrep * U_mat_N)[1:2, 1:2]
-            end
+        for ν = 1:5
+            Q̃_mat[:, :, ν, atom] = Hermitian(U_mat' * Q_mat_N[ν] * U_mat)
         end
+        T̃_mat[:, :, atom] = Hermitian(U_mat' * sys.interactions_union[atom].aniso.matrep * U_mat)
     end
 
     return s̃_mat, T̃_mat, Q̃_mat
 end
 
-function SpinWaveTheory(sys::System{N}, energy_ϵ::Float64=1e-8, energy_tol::Float64=1e-6) where N
-    (N == 0) && error("`SpinWaveTheory` requires an SU(N) `System`. Dipole mode is not currently supported.") 
-
+function SpinWaveTheory(sys::System{N}; energy_ϵ::Float64=1e-8, energy_tol::Float64=1e-6) where N
     # Reshape into single unit cell
     cellsize_mag = cell_dimensions(sys) * diagm(collect(sys.latsize))
     sys = reshape_geometry_aux(sys, (1,1,1), cellsize_mag)
 
-    s̃_mat, T̃_mat, Q̃_mat = generate_local_sun_gens(sys)
+    # Computes the Stevens operator in the local reference frame and the SO(3) rotation matrix from global to local frame
+    # (:dipole mode only)
+    c′_coef = Vector{StevensExpansion}()
+    R_mat   = Vector{Mat3}()
+    if sys.mode == :SUN
+        s̃_mat, T̃_mat, Q̃_mat = generate_local_sun_gens(sys)
+    elseif sys.mode == :dipole
+        Nₘ, Ns = length(sys.dipoles), sys.Ns[1] # number of magnetic atoms and dimension of Hilbert 
+        s_mat_N = spin_matrices(Ns)
+        Nₘ = length(sys.dipoles)
+        s̃_mat = Array{ComplexF64, 4}(undef, 2, 2, 3, Nₘ)
+        T̃_mat = zeros(ComplexF64, 0, 0, 0)
+        Q̃_mat = zeros(ComplexF64, 0, 0, 0, 0)
+        Nₘ = length(sys.dipoles)
+        R  = zeros(Float64, 3, 3)
+        for atom = 1:Nₘ
+            θ, ϕ = dipole_to_angles(sys.dipoles[1, 1, 1, atom])
+            # U_mat = exp(-1im*ϕ*s_mat_N[3]) * exp(-1im*θ*s_mat_N[2])
+            # SO(3) rotation that aligns the quantization axis. Important: since we will project out bosons that correspond to multipolar fluctuations,
```

Comment:
I am curious to know more about this. How will the convention for R become important?

Reply:
Here we want to ensure that the operators' second row/column in the local reference frame corresponds to the boson with `Delta S = 1`. Then we can project out the bosons corresponding to multipolar fluctuations for the exchange interaction. Doing so allows us to use the same code when constructing the spin wave Hamiltonian for the exchange interactions.

Reply:
What specifically was the thing you had to change to get it to behave the way you want?

Reply:
I asked Hao about this in person, and I think there are some subtleties. Maybe we should all have a conversation together.